system/modules/core/library/Contao/Validator.php

```diff
@@ -388,4 +388,35 @@ public static function isInsecurePath($strPath)
 
 		return false;
 	}
+
+
+	/**
+	 * Valid file name
+	 *
+	 * @param mixed $strName The file name
+	 *
+	 * @return boolean True if the file name is valid
+	 */
+	public static function isValidFileName($strName)
+	{
+		// Special characters not supported on e.g. Windows
+		if (preg_match('@[\\\\/:*?"<>\|]@', $strName))
+		{
+			return false;
+		}
+
+		// Invisible control characters or unused code points
+		if (preg_match('/[\pC]/u', $strName))
+		{
+			return false;
+		}
+
+		// Must not be longer than 255 characters
+		if (utf8_strlen($strName) > 255)
```

Comment:
409 doesn’t check if it is valid UTF-8, in fact it can be bypassed with invalid UTF-8:

```php
preg_match('/[\pC]/u', "normal string") === 0
preg_match('/[\pC]/u', "string with null byte \0") === 1
preg_match('/[\pC]/u', "string with null byte \0 and invalid UTF-8 \xC3") === false
```

So we should change the check to `if (preg_match('/[\pC]/u', $strName) !== 0)`.